### PR TITLE
feat: validate platforms

### PR DESCRIPTION
**Pull request by tjallingt**

### CODE REVIEW

File: lib/App/index.js [hunk 1]
```diff
@@ -209,7 +209,7 @@ class App {
           const isAppCapability = (typeof appJson.capabilities !== 'undefined' && typeof appJson.capabilities[capabilityId] !== 'undefined');
 
           if (!isSystemCapability && !isAppCapability) {
-            throw new Error(`Invalid capability: ${capability}`);
+            throw new Error(`drivers.${driver.id} invalid capability: ${capability}`);
```
Comment (tjallingt) [inline lib/App/index.js:212]: Couldn't figure out a message style I was completely happy with.
This change breaks with the consistent "Invalid ..." elsewhere but does include the driver id which I think is important

File: lib/App/index.js [hunk 2]
```diff
@@ -285,10 +285,29 @@ class App {
           await this._validateImages(driver.images, 'driver');
         }
 
+        // validate `appJson.drivers[].connectivity`
```
Comment (WeeJeWel) [inline lib/App/index.js:288]: This comment seems to be too high

File: lib/App/index.js [hunk 3]
```diff
@@ -285,10 +285,29 @@ class App {
...
+        if (levelVerified) {
```
Comment (WeeJeWel) [inline lib/App/index.js:289]: This check should always happen, regardless of the validation level.

File: lib/App/index.js [hunk 4]
```diff
@@ -285,10 +285,29 @@ class App {
...
+          if (driver.platforms.includes('local')) {
+            if (appJson.platforms.includes('local') === false) {
+              throw new Error(`drivers.${driver.id} invalid 'platforms': App manifest does not list 'local' as a supported platform.`);
+            }
+          }
+
+          if (driver.platforms.includes('cloud')) {
```
Comment (WeeJeWel) [inline lib/App/index.js:296]: Also copy this check to Flow cards

File: lib/App/index.js [hunk 5]
```diff
@@ -285,10 +285,29 @@ class App {
...
+            if (appJson.platforms.includes('cloud') === false) {
+              throw new Error(`drivers.${driver.id} invalid 'platforms': App manifest does not list 'cloud' as a supported platform.`);
+            }
+
+            if (driver.connectivity.includes('lan') || driver.connectivity.includes('rf868')) {
+              throw new Error(`drivers.${driver.id} invalid 'connectivity': Homey Bridge does not support 'lan' or 'rf868'.`);
```
Comment (WeeJeWel) [inline lib/App/index.js:302]: Replace `Homey Bridge` with `Platform 'cloud'`

**Review by WeeJeWel — CHANGES_REQUESTED** [review]

File: lib/App/index.js (hunk 2 above)
Comment (tjallingt) [reply lib/App/index.js]: Duplicated some of the checks so that the code is clearer and the comments make more sense 👍

**Review by jeroenwienk — APPROVED** [review]

**Review by WeeJeWel — APPROVED** [review]

File: lib/App/index.js (hunk 1 above)
Comment (RobinBol) [reply lib/App/index.js:212]: Invalid capability "${capability}" for driver "${driver.id}"

File: lib/App/index.js [hunk 6]
```diff
@@ -285,10 +285,31 @@ class App {
           await this._validateImages(driver.images, 'driver');
         }
 
+        // validate `appJson.drivers[].platforms`
+        if (driver.platforms && appJson.platforms) {
```
Comment (RobinBol) [inline lib/App/index.js:289]: Do we need to check if `driver.platforms` is an array before using `includes()`? e.g. throw "driver.platforms must be an array"

Comment (RobinBol) [reply lib/App/index.js:289]: (same for connectivity perhaps)

**Pull request merged**